src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -23,41 +24,68 @@
    [schema.core :as s]
    [toucan2.core :as t2])
   (:import
-   (java.sql Connection PreparedStatement)))
+   (java.sql Connection PreparedStatement SQLException)))
 
 (set! *warn-on-reflection* true)
 
-(defmulti parse-sql-error
-  "Parses the raw error message returned after an error in the driver database occurs, and converts it into a sequence
-  of maps with a :column and :message key indicating what went wrong."
-  {:arglists '([driver database e]), :added "0.44.0"}
-  driver/dispatch-on-initialized-driver
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                               Error handling                                                   |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defmulti maybe-parse-sql-error
+  "Try to parse the error message returned by JDBC driver.
+
+  The methods should returns a map of:
+  - type: the error type. Check [[metabase.actions.error]] for the full list
+  - message: a nice message summarized of what went wrong
+  - errors: a map from field-name => sepcific error message. This is used by UI to display per fields error."
+  {:arglists '([driver error-type database error-message]), :added "0.48.0"}
+  (fn [driver error-type _database _error-message]
+   [(driver/dispatch-on-initialized-driver driver) error-type])
   :hierarchy #'driver/hierarchy)
 
-(defn- parse-error
-  "Returns errors in a way that indicates which column had the problem. Can be used to highlight errors in forms."
+(defmethod maybe-parse-sql-error :default
+  [_driver _error-type _database _e]
+  nil)
+
+(defn- parse-sql-error
   [driver database e]
-  (let [default-method? (= (get-method parse-sql-error driver)
-                           (get-method parse-sql-error :default))]
-    (if default-method?
```

Comment:
this logic was also the motivation for the refactor, it's really hard to know what will be parsed upstream and what's being processed.